assembly/src/assembler/instruction/mem_ops.rs

```diff
@@ -1,34 +1,43 @@
 use super::{
-    push_felt, push_u16_value, AssemblerError, AssemblyContext, CodeBlock, Felt, Operation::*,
-    SpanBuilder, StarkField,
+    push_u16_value, push_u32_value, validate_param, AssemblerError, AssemblyContext, CodeBlock,
+    Felt, Operation::*, SpanBuilder, StarkField,
 };
 
 // INSTRUCTION PARSERS
 // ================================================================================================
 
-/// Appends operations to the span block to execute a memory read operation. This includes reading
-/// a single element or an entire word from either local or global memory. Specifically, this
-/// handles mem_load, mem_loadw, loc_load, and loc_loadw instructions.
+/// Appends operations to the span needed to execute a memory read instruction. This includes
+/// reading a single element or an entire word from either local or global memory. Specifically,
+/// this handles mem_load, mem_loadw, loc_load, and loc_loadw instructions.
 ///
 /// VM cycles per operation:
 /// - mem_load(w): 1 cycle
 /// - mem_load(w).b: 2 cycles
 /// - loc_load(w).b:
 ///    - 4 cycles if b = 1
 ///    - 3 cycles if b != 1
+///
+/// # Errors
+/// Returns an error if we are reading from local memory and local memory index is greater than
+/// the number of procedure locals.
 pub fn mem_read(
     span: &mut SpanBuilder,
     context: &AssemblyContext,
-    imm: Option<&Felt>,
+    addr: Option<&Felt>,
     is_local: bool,
     is_single: bool,
 ) -> Result<Option<CodeBlock>, AssemblerError> {
-    let num_proc_locals = context.num_proc_locals();
-    match imm {
-        Some(imm) if is_local => push_local_addr(span, imm, num_proc_locals)?,
-        None if is_local => unreachable!("local always contains imm value"),
-        Some(imm) => push_felt(span, *imm),
-        None => (),
+    // if the address was provided as an immediate value, put it onto the stack
+    if let Some(addr) = addr {
```

Comment:
non-blocking nit: match pattern might be more canonical/ergonomic to rust since it produces a concise description of the logic that is being branched

Reply:
Yes, would be great to update this in the future clean-ups.